### DISCUSSION TITLE
Does Mulvus support multi-col vector search for now?

### DISCUSSION THREAD

**Discussion by JackTan25** [milvus-io/milvus#28361] [Q&A and General discussion]
there is a similar issue here #12085 in 2021.12, but I don't know the progress for now. Does this support for now?

**Comment by xiaofan-luan** [comment]
We are gonna to support multi vector features in 2.4 

The API design is like

curl --request POST \
     --url "${MILVUS_HOST}:${MILVUS_PORT}/v2/vector/search" \
     --header "Authorization: Bearer ${TOKEN}" \
     --header "accept: application/json" \
     --header "content-type: application/json" \
     -d '{<request_body>}'
# request_body for multi-vector search
-- simple version & phase 1 --
 {
    "collectionName": "book",
    "search": {
            "search_by": { # get the best 100 book_intro_vectors that has price > 5
                "field": "book_intro_vector",
                "vector": [1, 2, ...], 
                # "limit": default # = root(n) * limit
            },
            "search_by": {
                "field": "book_cover_vector",
                "vector": [2, 3, ...],
                #"limit": default # = root(n) * limit
            },
    },
    "rerank": {
        "strategy": "rrf", 
    },
    "limit": 10,
    # "output_fields" : ["id", "score"],
}

 -- detailed version & phase 2 --
  {
    # step 1: search input
    "collectionName": "book",
    "search": [
            "search_by": { # get the best 100 book_intro_vectors that has price > 5
                "vector_field": "book_intro_vector",
                "data": [1, 2, ...], 
                "limit": 100 # = root(n) * limit
                #"norm_type": minmax # this is how you measure score for this way callback
            },
            "search_by": {
                "vector_field": "book_cover_vecor",
                "data": [2, 3, ...],
                "filter": "price > 10"
                "limit": 200 # = root(n) * limit
                #"norm_type": minmax # this is how you measure score for this way callback
            },
            ## May Open for future scalar search
            #"search_by": { # search by non-vector field, key word match
            #    "filter": "price > 5"
            #    "limit": 300 # = root(n) * limit
            #},
    ],
    # step 2: rerank strategy
    # 2.1 rrf algorithm
    "rerank": {
        "strategy": "rrf", # rrf is a general method, donot need score of search 
        "parameter": {
            "k": 60 # k is a parameter for rrf
    },
    # 2.2 average weighted
    "rerank": {
        "strategy": "avg", # avg is customized method, need normolization to define score for each search_by
        "parameter": {
            "required_input_field": [book_intro_vector, book_cover_vecor]
            "weight":[0.1, 0.4], # which means the parameter for score for 2-way call is 1/5 plus 4/5
         }
    },
    # step 3: define output fields and over all
    "limit": 150
    "output_fields" : ["id", "url", "title"],
    #"consistency_level": "Strong" 
}


-- future ranker & phase 3 --
    {
    # for future ranker
    "rerank": {
        "strategy": "udf", # 自定义表达式
        "parameter": {
            "expression":{"0.6*book_intro_vector+0.2*book_cover_vecor"} # just an example, need further discussion
        }
    },
    "rerank": {
        "strategy": "model", # customized reranking
        "parameter": {
            "url":{"model_url"}
        }
    },

**Comment by xiaofan-luan** [comment]
generally speaking, we will retrieve on multiple columns and reranking based on RRF or hybrid scoreing
